packages/contracts-bedrock/contracts/universal/Proxy.sol

```diff
@@ -0,0 +1,155 @@
+
+    /***************
+     * Constructor *
+     ***************/
+
+    /**
+     * @param _owner Address of the initial contract owner.
+     */
+    constructor(address _owner) {
+        _setOwner(_owner);
+    }
+
+    // slither-disable-next-line locked-ether
+    fallback() external payable {
+        // Proxy call by default.
+        _doProxyCall();
+    }
+
+    modifier onlyOwner() {
+        address owner = _getOwner();
+        require(msg.sender == owner);
+        _;
+    }
+
+    /**********************
+     * External Functions *
+     **********************/
+
+    /**
+     * @param _implementation The address of the implementation contract
+     */
+    function setImplementation(address _implementation) external onlyOwner {
+        _setImplementation(_implementation);
+    }
+
+    /**
+     * Changes the owner of the proxy contract. Only callable by the owner.
+     * @param _owner New owner of the proxy contract.
+     */
+    // slither-disable-next-line external-function
+    function setOwner(address _owner) external onlyOwner {
+        _setOwner(_owner);
+    }
+
+    /**
+     * Queries the owner of the proxy contract. Can only be called by the owner OR by making an
+     * eth_call and setting the "from" address to address(0).
+     * @return Owner address.
+     */
+    // slither-disable-next-line external-function
+    function getOwner() external view returns (address) {
+        return _getOwner();
+    }
+
+    /**
+     * Queries the implementation address. Can only be called by the owner OR by making an
+     * eth_call and setting the "from" address to address(0).
+     * @return Implementation address.
+     */
+    // slither-disable-next-line external-function
+    function getImplementation() external view returns (address) {
+        return _getImplementation();
+    }
+
+    /**********************
+     * Internal Functions *
+     **********************/
+
+    /**
+     * Sets the implementation address.
+     * @param _implementation New implementation address.
+     */
+    function _setImplementation(address _implementation) internal {
+        assembly {
+            sstore(IMPLEMENTATION_KEY, _implementation)
+        }
+    }
+
+    /**
+     * Queries the implementation address.
+     * @return implementation address.
+     */
+    function _getImplementation() internal view returns (address implementation) {
+        assembly {
+            implementation := sload(IMPLEMENTATION_KEY)
+        }
+    }
+
+    /**
+     * Changes the owner of the proxy contract.
+     * @param _owner New owner of the proxy contract.
+     */
+    function _setOwner(address _owner) internal {
+        assembly {
+            sstore(OWNER_KEY, _owner)
+        }
+    }
+
+    /**
+     * Queries the owner of the proxy contract.
+     * @return owner address.
+     */
+    function _getOwner() internal view returns (address owner) {
+        assembly {
+            owner := sload(OWNER_KEY)
+        }
+    }
```

Comment:
I think we've generally agreed to avoid named return arguments. In this case though it's much nice than this, so I'd be OK to let this in if @smartcontracts is.

```solidity
    function _getOwner() internal view returns (address) {
		address owner;
        assembly {
            owner := sload(OWNER_KEY)
        }
        return owner;
    }
```

Reply:
Although I agree it's nicer here, let's use named returns for consistency. I'd like to be able to install a "no named returns" linter rule in the future.

Reply:
You mean:
> let's _not_ use named returns for consistency
👍